text/0000-extended-standard-library.md

```diff
@@ -0,0 +1,370 @@
+
+## Naming and Namespaces
+
+Earlier it was stated that all crates in the ESL would ideally have names that:
+
+1) Clearly describe the functionality they offer
+2) Contain a reliable indicator of ESL membership
+
+For practical reasons this may not be the case from the start of the ESL, but it's something to think about working towards in the future.
+
+### Clearly Descriptive Crate Names
+
+Because namespacing for crates is not implemented, crate names, rather than source entity names, must be unique (e.g. "CreativeName" vs. "CreativeOrg::DescriptiveName"). This encourages creative naming in order to achieve uniqueness. As a result, crates coming into ESL are likely to have names that are creative rather than clearly descriptive.
+
+Clearly descriptive crate names would help Rust developers find what they need more easily. While crates can have descriptions associated with them, reading descriptions is slower than reading names.
+
+Perhaps more importantly though, it would also help with the readability of Rust code since crate names are used in Rust code frequently. It's the difference between "creative::ClientConfig" and "tls::clientConfig". This can be worked around with "use [creative] as [descriptive]" but that is more work and not always done in clear ways.
+
+### Reliably Indicating ESL Membership
+
+Lack of namespacing also make it difficult to clearly and reliably indicate membership in the ESL, where reliably means that only crates in the ESL can indicate membership in the ESL.
+
+As part of understanding and mitigating supply chain risks, Rust developers should be able to easily tell which crates come from the same source. Unfortunately, this is not easy today.
+
+Adding support for namespacing crates is the best way to improve the situation. [Rust RFC 3243](https://rust-lang.github.io/rfcs/3243-packages-as-optional-namespaces.html) has been passed but is not yet implemented.
+
+In the future, it would be good to implement RFC 3243 and get ESL crates into an ESL namespace (e.g. "esl::" or "ext_std::"). This would be an improvement over requiring GitHub organization checks, and it would allow ESL membership for dependencies to be reflected in Rust application code.
+
+## Support for no_std
+
+It should be possible, both technically and in terms of policy, for ESL crates to support no_std. This would not necessarily be feasible or make sense for all ESL crates, but it might be something to consider for some.
+
+# Appendix A
+[appendix-a]: #appendix-a
+
+## Example Policies
+
+A simplified list of example policies to use as potential starting point.
+
+### Evolving APIs / Breaking Changes
+
+ESL crates will be versioned with [Semantic Versioning](https://semver.org/).
+
+Patch and minor updates may happen at any time.
+
+Major updates will happen no more than once per year unless necessitated by a significant security concern.
```

Comment:
Not to debate policy but just so I don't forget, we should consider a little more nuance here.

Problems with major versions
- Work to upgrade
- Interop when in a public API of another package
- Compile times from multiple copies

Not everything is big and not everything is meant to be in a public API.  As for being disruptive, we should probably encourage (but not require) breaking releases to be removal of deprecated functionality as discussed on [matklad's blog](https://matklad.github.io/2024/11/23/semver-is-not-about-you.html#SemVer--) and as I try to do with clap and winnow as much as possible.

Reply:
I don't think I understand what you're trying to get at here. When you say "major" do you mean the first number in the three number sequence? cargo's vision of versions is to treat the first non-zero version as the "breaking change" number, and the next numbers as compatible, right? So how is `0.x.y` somehow better than `x.y.z`? An incompatible version is just incompatible.

Reply:
The topic I'm replying to and my comment and "major" to refer to the relative position in the version number and not the absolute position (wish we had established names for these)

My comment is speaking to how to manage breaking releases.

Reply:
I think I saw where some of the confusion came from and made an edit that  I hope helps.